testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java

```diff
@@ -99,31 +107,77 @@ public void testGetAll() {
     }
 
     @Test
-    public void testGetByDomain() {
-        // create some organizations with a domain already set.
-        for (int i = 0; i < 5; i++) {
-            createOrganization("test-org-" + i, "testorg" + i + ".org");
-        }
-
-        // search for an organization with an existing domain.
-        List<OrganizationRepresentation> existing = testRealm().organizations().getAll("testorg2.org");
-        assertEquals(1, existing.size());
+    public void testSearch() {
+        // create some organizations with different names and domains.
+        createOrganization("acme", "acme.org", "acme.net");
+        createOrganization("Gotham-Bank", "gtbank.com", "gtbank.net");
+        createOrganization("wayne-industries", "wayneind.com", "wayneind-gotham.com");
+        createOrganization("TheWave", "the-wave.br");
+
+        // test exact search by name (e.g. 'wayne-industries'), e-mail (e.g. 'gtbank.net'), and no result (e.g. 'nonexistent.com')
+        List<OrganizationRepresentation> existing = testRealm().organizations().search("wayne-industries", true, 0, 10);
+        assertThat(existing, hasSize(1));
         OrganizationRepresentation orgRep = existing.get(0);
-        assertEquals("test-org-2", orgRep.getName());
-        assertEquals(1, orgRep.getDomains().size());
-        OrganizationDomainRepresentation domainRep = orgRep.getDomains().iterator().next();
-        assertEquals("testorg2.org", domainRep.getName());
-        assertFalse(domainRep.isVerified());
-
-        // search for an organization with an non-existent domain.
-        existing = testRealm().organizations().getAll("someother.org");
-        assertEquals(0, existing.size());
+        assertThat(orgRep.getName(), is(equalTo("wayne-industries")));
+        assertThat(orgRep.getDomains(), hasSize(2));
+        assertThat(orgRep.getDomain("wayneind.com"), not(nullValue()));
+        assertThat(orgRep.getDomain("wayneind-gotham.com"), not(nullValue()));
+
+        existing = testRealm().organizations().search("gtbank.net", true, 0, 10);
+        assertThat(existing, hasSize(1));
+        orgRep = existing.get(0);
+        assertThat(orgRep.getName(), is(equalTo("Gotham-Bank")));
+        assertThat(orgRep.getDomains(), hasSize(2));
+        assertThat(orgRep.getDomain("gtbank.com"), not(nullValue()));
+        assertThat(orgRep.getDomain("gtbank.net"), not(nullValue()));
+
+        existing = testRealm().organizations().search("nonexistent.org", true, 0, 10);
+        assertThat(existing, is(empty()));
+
+        // partial search matching name (e.g. 'wa' matching 'wayne-industries', and 'TheWave')
+        existing = testRealm().organizations().search("wa", false, 0, 10);
+        assertThat(existing, hasSize(2));
+        List<String> orgNames = existing.stream().map(OrganizationRepresentation::getName).collect(Collectors.toList());
+        assertThat(orgNames, containsInAnyOrder("wayne-industries", "TheWave"));
```

Comment:
yeah, I also thought the same and the test was actually testing the order, but when the tests ran using multiple DBs they failed because the behavior is not consistent among DBs. In some of them, the order by was case insensitive (so, words starting with capital letters would be mixed with lower-cased words), while for others it was sensitive (so words starting with capital letters would come first, then all words starting with lower case letters, etc).

Looks like default collation is different among DBs, so I didn't want to make the test dependent on a specific collation. One way to fix this would be to `ORDER BY lower(name)`, which would consistently return a case insensitive order, but it looks like there's some performance penalties when doing this for big result sets.

Reply:
By the way, we might need to revisit this for all entities at some point because we have several queries with `ORDER BY` in code. In some cases, a `sort(comparator)` is used in the returned stream, but that doesn't make results consistent among different collations when you have pagination.

If the query fetches only the first 10 results in a paginated search, then in a DB with case-insensitive `ORDER BY` you will have mixed-case results. In a different DB the first 10 objects might all start with capital letters, so ordering after the results are returned doesn't seem to make a lot of sense as it doesn't guarantee consistent results.